### PR TITLE
Feature/fix approval handling

### PR DESCRIPTION
**Pull request by matthias-geuchen**
### Reason for this PR
- when adding new matriculation data, the contract should check if the existing approvals suffice

### Changes in this PR
- addmatriculationData now checks approvals
- add groups to approvals
- update tests
- add submission result
- add AccessManager for required approvals

### CODE REVIEW

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/ApprovalList.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        return Objects.hash(users, groups);
+    }
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class MatriculationData {\n");
+
+        sb.append("    users: ").append(toIndentedString(users)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/ApprovalList.java:100]:
```suggestion
        sb.append("class MatriculationData {\n");
        sb.append("    users: ").append(toIndentedString(users)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/MatriculationData.java [hunk 2]
```diff
@@ -1,135 +1,129 @@
...
-    if (o == null) {
-      return "null";
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class MatriculationData {\n");
+
+        sb.append("    enrollmentId: ").append(toIndentedString(enrollmentId)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/MatriculationData.java:88]:
```suggestion
        sb.append("class MatriculationData {\n");
        sb.append("    enrollmentId: ").append(toIndentedString(enrollmentId)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/SubjectMatriculation.java [hunk 3]
```diff
@@ -1,116 +1,105 @@
...
+        return Objects.hash(fieldOfStudy, semesters);
+    }
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class SubjectImmatriculationInterval {\n");
+
+        sb.append("    fieldOfStudy: ").append(toIndentedString(fieldOfStudy)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/SubjectMatriculation.java:87]:
```suggestion
        sb.append("class SubjectImmatriculationInterval {\n");
        sb.append("    fieldOfStudy: ").append(toIndentedString(fieldOfStudy)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/DetailedError.java [hunk 4]
```diff
@@ -1,133 +1,136 @@
...
+        return Objects.hash(type, title, invalidParams);
+    }
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class DetailedError {\n");
+
+        sb.append("    type: ").append(toIndentedString(type)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/DetailedError.java:117]:
```suggestion
        sb.append("class DetailedError {\n");
        sb.append("    type: ").append(toIndentedString(type)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/GenericError.java [hunk 5]
```diff
@@ -11,91 +11,92 @@
...
-    if (o == null) {
-      return "null";
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class DetailedError {\n");
+
+        sb.append("    type: ").append(toIndentedString(type)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/GenericError.java:84]:
```suggestion
        sb.append("class DetailedError {\n");
        sb.append("    type: ").append(toIndentedString(type)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/InvalidParameter.java [hunk 6]
```diff
@@ -1,100 +1,102 @@
...
+        return Objects.hash(name, reason);
+    }
+
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class InvalidParameter {\n");
+
+        sb.append("    name: ").append(toIndentedString(name)).append("\n");
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/model/errors/InvalidParameter.java:84]:
```suggestion
        sb.append("class InvalidParameter {\n");
        sb.append("    name: ").append(toIndentedString(name)).append("\n");
```

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/helper/GsonWrapper.java [hunk 7]
```diff
@@ -13,7 +13,7 @@
 
 public class GsonWrapper {
 
-    private static final Gson gson = new GsonBuilder()
+    private static final Gson gson = new GsonBuilder().disableHtmlEscaping()
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/util/helper/GsonWrapper.java:16]:
```suggestion
    private static final Gson gson = new GsonBuilder()
```

**Review by NikoBergemann — CHANGES_REQUESTED** [review]
changes as commented in single comments

**Review by mco33415 — APPROVED** [review]

**Review by NikoBergemann — APPROVED** [review]

**Pull request merged**